espr/src/ir/constraints.rs

```diff
@@ -1,383 +1,193 @@
-    type Output = Self;
-    fn sub(self, q: PartialComplexEntity) -> Self {
-        Constraint {
-            parts: self.parts.into_iter().filter(|p| p != &q).collect(),
-        }
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn constraint_oneof() {
+        let st = ast::SyntaxTree::parse(
+            r#"
+            SCHEMA test_schema;
+              ENTITY base SUPERTYPE OF (ONEOF (sub1, sub2));
+                x: REAL;
+              END_ENTITY;
+
+              ENTITY sub1 SUBTYPE OF (base);
+                y1: REAL;
+              END_ENTITY;
+
+              ENTITY sub2 SUBTYPE OF (base);
+                y2: REAL;
+              END_ENTITY;
+            END_SCHEMA;
+            "#,
+        )
+        .unwrap();
+
+        let ns = Namespace::new(&st);
+        let c = Constraints::new(&ns, &st).unwrap();
+
+        let scope = Scope::root().schema("test_schema");
+        assert_eq!(
+            c,
+            Constraints {
+                instantiables: maplit::hashmap! {
+                    Path::entity(&scope, "base") => vec![
+                        vec![Path::entity(&scope, "sub1")],
+                        vec![Path::entity(&scope, "sub2")],
+                    ]
+                }
+            }
+        );
     }
-}
 
-// [A, A & B, A & C, A & B & D, B & C, D]/[B, D] ≡ [A & B, A & B & D, B & C, D]
-impl std::ops::Div for Constraint {
-    type Output = Self;
-    fn div(self, rhs: Constraint) -> Self {
-        Constraint {
-            parts: self
-                .parts
-                .into_iter()
-                .filter(|p| {
-                    for q in &rhs.parts {
-                        if q.indices.iter().all(|j| p.indices.binary_search(j).is_ok()) {
-                            return true;
-                        }
-                    }
-                    false
-                })
-                .collect(),
-        }
+    #[test]
+    fn constraint_andor() {
+        // Based on `ANDOR` example in ISO-10303-11
+        let st = ast::SyntaxTree::parse(
+            r#"
+            SCHEMA test_schema;
+              ENTITY person SUPERTYPE OF (employee ANDOR student);
+              END_ENTITY;
+              ENTITY employee SUBTYPE OF (person);
+              END_ENTITY;
+              ENTITY student SUBTYPE OF (person);
+              END_ENTITY;
+            END_SCHEMA;
+            "#,
+        )
+        .unwrap();
+
+        let ns = Namespace::new(&st);
+        let c = Constraints::new(&ns, &st).unwrap();
+
+        let scope = Scope::root().schema("test_schema");
+        assert_eq!(
+            c,
+            Constraints {
+                instantiables: maplit::hashmap! {
+                    Path::entity(&scope, "person") => vec![
+                        vec![Path::entity(&scope, "employee")],
+                        vec![Path::entity(&scope, "employee"), Path::entity(&scope, "student")],
+                        vec![Path::entity(&scope, "student")],
```

Comment:
This order should be following:

```suggestion
                        vec![Path::entity(&scope, "employee")],
                        vec![Path::entity(&scope, "student")],
                         vec![Path::entity(&scope, "employee"), Path::entity(&scope, "student")],
```

Reply:
This is due to `Vec` uses lexicographical order https://doc.rust-lang.org/std/cmp/trait.Ord.html#lexicographical-comparison